javalib/src/main/scala/java/util/concurrent/ConcurrentLinkedQueue.scala

```diff
@@ -0,0 +1,153 @@
+package java.util.concurrent
+
+import java.util._
+import scala.collection.JavaConversions._
+
+class ConcurrentLinkedQueue[E]()
+    extends AbstractQueue[E] with Queue[E] with Serializable {
+
+  def this(c: Collection[_ <: E]) = {
+    this()
+    addAll(c)
+  }
+
+  import ConcurrentLinkedQueue._
+
+  private var head: Node[E] = null
+  private var last: Node[E] = null
+
+  private var _size: Double = 0
+
+  override def add(e: E): Boolean =
+    if (e == null) throw new NullPointerException()
+    else {
+      val oldLast = last
+
+      last = new Node(e)
+
+      _size += 1
+
+      if (oldLast ne null)
+        oldLast.next = last
+      else
+        head = last
+
+      true
+    }
+
+  override def offer(e: E): Boolean =
+    add(e)
+
+  override def poll(): E = {
+    if (isEmpty()) null.asInstanceOf[E]
+    else {
+      val oldHead = head
+      head = oldHead.next
+
+      if (head eq null)
+        last = null
+
+      _size -= 1
+      oldHead.value
+    }
+  }
+
+  override def peek(): E =
+    if (isEmpty()) null.asInstanceOf[E]
+    else head.value
+
+  override def isEmpty(): Boolean =
+    _size == 0
+
+  override def size(): Int =
+    _size.toInt
+
+  private def getNodeAt(index: Int): Node[E] = {
+    if (index == 0) head
+    else {
+      var current: Node[E] = null
+      current = head
+      for (_ <- 0 until index)
+        current = current.next
+      current
+    }
+  }
+
+  private def removeNode(node: Node[E]): E = {
+    if (node eq head) poll()
+    else {
```

Comment:
Same:
```scala
if (node eq head) {
  poll()
} else {
```
Also, the else branch will crash if the list is empty. It can happen with the following test case:
```scala
val q = new ConcurrentLinkedQueue
q.add(1)
val iter = q.iterator()
iter.next()
q.poll()
iter.remove() // must not throw
```
So this else should be:
```scala
} else if (head ne null) {
```
(if `head eq null`, there is nothing to do)